### ISSUE TITLE
docker stop correlated with temporary network connectivity loss

### ISSUE THREAD

**Issue by stephenliang** [docker/docker#14148]
I'm currently running docker containers controlled by Mesos via the Mesos Jenkins plugin. In short, the idea is to start Jenkins slave docker containers as they are needed. Occassionally, I see Jenkins slaves go down due to "Connection reset by peer". I see no fault on the Jenkins side, though it is not out of the question.

The reason why I'm reporting this as a docker issue instead of a Jenkins issue is that each time there is a "Connection reset by peer", all my containers temporarily lose network connectivity. I can prove this by launching another docker container and running `nc` on it every single second. During a "connection reset by peer" on a different container, I can see this:

        Tue Jun 23 21:15:21 CDT 2015
        Connection to 10.228.34.164 7778 port [tcp/interwise] succeeded!
        Tue Jun 23 21:15:22 CDT 2015
        nc: connect to 10.228.34.164 port 7778 (tcp) failed: Connection timed out
        Tue Jun 23 21:15:31 CDT 2015
        Connection to 10.228.34.164 7778 port [tcp/interwise] succeeded!

Each time this happens, I lose network on docker containers (not the host) for roughly 30 seconds to a minute. The way Jenkins slave connects to a remote is that it starts a java application in the docker container, and then it connects to a well known address and port on the jenkins master. If this port is closed unexpectedly, we can get the connection reset by peer. However, I don't see anything in the Jenkins slave logs nor on the master that indicates that anything was about to be closed. In fact, Jenkins was still running mid-way.

In addition, I can see that during this time, docker is almost always stopping containers, as such in `/var/log/messages`: https://gist.github.com/stephenliang/de6392a080b5fc3d00c9

Here's another example, different time but similar circumstances (with NetworkManager turned off): https://gist.github.com/stephenliang/1c931bbabd1408132e4b

I have tried various versions of Docker, starting with 1.4, 1.5, 1.6, and 1.7. All are reproducible. I have also tried with the mainline kernel, and with NetworkManager disabled.

I'm unable to reproduce this by simulating a `docker kill` followed by a `docker stop` or vice versa. This only seems to be happening during a Jenkins build.

Reproducible Environment 1:
OS: Centos 7.1
uname: Linux mesos-2.example.net 3.10.0-229.4.2.el7.x86_64 #1 SMP Wed May 13 10:06:09 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
Docker Version:

        Client version: 1.7.0
        Client API version: 1.19
        Go version (client): go1.4.2
        Git commit (client): 0baf609
        OS/Arch (client): linux/amd64
        Server version: 1.7.0
        Server API version: 1.19
        Go version (server): go1.4.2
        Git commit (server): 0baf609
        OS/Arch (server): linux/amd64

Docker Info:

        Containers: 8
        Images: 33
        Storage Driver: devicemapper
         Pool Name: docker-253:0-537952437-pool
         Pool Blocksize: 65.54 kB
         Backing Filesystem: xfs
         Data file: /dev/loop0
         Metadata file: /dev/loop1
         Data Space Used: 31.76 GB
         Data Space Total: 107.4 GB
         Data Space Available: 75.61 GB
         Metadata Space Used: 18.94 MB
         Metadata Space Total: 2.147 GB
         Metadata Space Available: 2.129 GB
         Udev Sync Supported: true
         Deferred Removal Enabled: false
         Data loop file: /var/lib/docker/devicemapper/devicemapper/data
         Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
         Library Version: 1.02.93-RHEL7 (2015-01-28)
        Execution Driver: native-0.2
        Logging Driver: json-file
        Kernel Version: 3.10.0-229.4.2.el7.x86_64
        Operating System: CentOS Linux 7 (Core)
        CPUs: 24
        Total Memory: 47 GiB
        Name: mesos-2.example.net
        ID: JKJT:JZOL:74SD:O473:XHBF:KBTA:UYWN:OV4E:BMWY:EUZZ:5EBU:IEUV

Reproducible Environment 2:

uname: Linux mesos-3.example.net 4.0.5-1.el7.elrepo.x86_64 #1 SMP Sun Jun 7 08:52:16 EDT 2015 x86_64 x86_64 x86_64 GNU/Linux

docker version: 

        Client version: 1.4.0
        Client API version: 1.16
        Go version (client): go1.3.3
        Git commit (client): 4595d4f
        OS/Arch (client): linux/amd64
        Server version: 1.4.0
        Server API version: 1.16
        Go version (server): go1.3.3
        Git commit (server): 4595d4f

docker info:

        Containers: 7
        Images: 308
        Storage Driver: btrfs
        Execution Driver: native-0.2
        Kernel Version: 4.0.5-1.el7.elrepo.x86_64
        Operating System: CentOS Linux 7 (Core)
        CPUs: 16
        Total Memory: 47.16 GiB
        Name: mesos-3.example.net
        ID: A5YC:PUEU:R7VE:BHTN:LVEI:JNTV:D4P6:KHP2:FVXL:KZTY:7YGN:TMPW
        Debug mode (server): false
        Debug mode (client): true
        Fds: 47
        Goroutines: 65
        EventsListeners: 0
        Init Path: /usr/bin/docker
        Docker Root Dir: /var/lib/docker

**Comment by stephenliang** [comment]
Also, the containers are set up this way:

Random port bound to port 9999, random port bound to port 5900 running a centos:6 based docker image.

**Comment by etuttle** [comment]
Today I saw a container fail an http health check for about 15 seconds right after another container was stopped.  I'm running Docker 1.6.2 / Centos6.6.  Might be the same issue!

**Comment by etuttle** [comment]
My health checks are no longer failing on `docker rm -f` after I applied a fix to the docker0 bridge, as suggested in https://github.com/docker/docker/issues/11407#issuecomment-97746220

**Comment by cpuguy83** [comment]
Is this still a problem on 1.8.2?

@etuttle centos 6 is no longer supported, and also looks like a different issue (especially since networking was re-written from scratch in 1.7).

**Comment by pgkelley4** [comment]
@cpuguy83 I am still seeing this issue on 1.8.2 and 1.9.0. The fix provided by @etuttle here: https://github.com/docker/docker/issues/11407#issuecomment-120172459 does fix it. 

It seems to happen when a container is stopped, but not every time. For example, I see network connectivity issues in some long running containers when doing:
```
#!/bin/bash

while true ; do
 docker run --rm ubuntu:14.04 /bin/echo 'Hello world'
 sleep 1
done
```

Docker info:
```
Containers: 33
Images: 153
Storage Driver: aufs
 Root Dir: /data/docker/aufs
 Backing Filesystem: extfs
 Dirs: 219
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.16.0-51-generic
Operating System: Ubuntu 14.04.1 LTS
CPUs: 8
Total Memory: 29.45 GiB
```

Docker version (although I also tried 1.9.0)
```Client:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:19:00 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   0a8c2e3
 Built:        Thu Sep 10 19:19:00 UTC 2015
 OS/Arch:      linux/amd64```

**Comment by stephenliang** [comment]
Just an update, we did apply the fix mentioned in https://github.com/docker/docker/issues/11407#issuecomment-97746220 and kept our Kernel at 4.0+. We have not seen this issue crop up in Docker 1.8 since we applied those workarounds.

**Comment by aboch** [comment]
In systems with kernel version `>=3.3` libnetwork explicitly sets the MAC address of the bridge interface after creating it. This would assure the system will not reassign the bridge MAC address to the lower MAC of the enslaving interfaces.

Because of a old bug on the kernel version check and another bug where the MAC setting was being done in the incorrect phase, above behavior is true only in docker versions `1.8.0` onward.

There is one thing to keep in mind though:
- If the `docker0` bridge is already present when starting docker `>=1.8.0`, libnetwork does not recreate the bridge. So the system is still vulnerable to the bridge MAC reassignment, if it was created by a docker with version `< 1.8.0`.

@stephenliang @pgkelley4
Can you confirm you still see the problem with docker >= 1.8.0 if you remove the `docker0` bridge interface on your system before starting the new daemon.

**Comment by aboch** [comment]
@stephenliang @pgkelley4
Another way to check whether your current `docker0` bridge is subject to the MAC address change, is to check the output of 
`cat /sys/class/net/docker0/addr_assign_type`
if you see `3` you are good, it means docker assigned the address and it will not change. If you  see `1` then anytime an interface is added/removed to the bridge, the bridge mac address could potentially change.

**Comment by stephenliang** [comment]
We hadn't seen any issues after clean installing CentOS 7.1 along with Kernel 4.0+ and Docker 1.8, however we did apply the workaround method I mentioned earlier.

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
